main.cpp

```diff
@@ -104,199 +114,474 @@ good to go!
-    std::cout << "---------------------\n" << std::endl; 
+    FloatType& subtract(float f );
+    FloatType& subtract(const DoubleType d );
+    FloatType& subtract(const IntType i);
     
-    // DoubleType/IntType object instanciation and method tests
-    // --------
-    std::cout << "Initial value of dt: " << dt.value << std::endl;
-    std::cout << "Initial value of it: " << it.value << std::endl;
-    // --------
-    std::cout << "Use of function concatenation (mixed type arguments) " << std::endl;
-    std::cout << "New value of dt = (dt * it) / 5.0f + ft = " << (dt.multiply(it).divide(5.0f).add(ft).value) << std::endl;
+    FloatType& multiply(float rhs );
+    FloatType& multiply(const DoubleType d );
+    FloatType& multiply(const IntType i);
+    
+    FloatType& divide(float rhs );
+    FloatType& divide(const DoubleType d );
+    FloatType& divide(const IntType i);
+};
+
+//-------------------------------------------------
+// DoubleType
+//-------------------------------------------------
+
+struct DoubleType
+{
+    double* value;
 
-    std::cout << "---------------------\n" << std::endl; 
+    DoubleType(double d) : value (new double{d}) {}
+    ~DoubleType()
+    {
+        //delete value;
+        value = nullptr;
+    }
+
+    DoubleType& add(double d );
+    DoubleType& add(const FloatType f);
+    DoubleType& add(const IntType f);
+
+    DoubleType& subtract(double d );
+    DoubleType& subtract(const FloatType f);
+    DoubleType& subtract(const IntType i);
+
+    DoubleType& multiply(double d );
+    DoubleType& multiply(const FloatType f);
+    DoubleType& multiply(const IntType i);
     
-    // Intercept division by 0
-    // --------
-    std::cout << "Intercept division by 0 " << std::endl;
-    std::cout << "New value of it = it / 0 = " << it.divide(0).value << std::endl;
-    std::cout << "New value of ft = ft / 0 = " << ft.divide(0).value << std::endl;
-    std::cout << "New value of dt = dt / 0 = " << dt.divide(0).value << std::endl;
+    DoubleType& divide(double d );
+    DoubleType& divide(const FloatType f);
+    DoubleType& divide(const IntType i);
+};
 
-    std::cout << "---------------------\n" << std::endl; 
 
-    std::cout << "good to go!\n";
+//-------------------------------------------------
+// IntType
+//-------------------------------------------------
 
-    return 0;
+struct IntType
+{
+    int* value;
+
+    IntType(int i) : value(new int{i}) {}
+    ~IntType()
+    {
+        //delete value;
+        value = nullptr;
+    }
+
+    IntType& add(int i );
+    IntType& add(const FloatType f );
+    IntType& add(const DoubleType d );
+
+    IntType& subtract(int i );
+    IntType& subtract(const FloatType f );
+    IntType& subtract(const DoubleType d );
+
+    IntType& multiply(int i );
+    IntType& multiply(const FloatType f );
+    IntType& multiply(const DoubleType d );
+
+    IntType& divide(int i );
+    IntType& divide(const FloatType f );
+    IntType& divide(const DoubleType d );
```

Comment:
you need to pass these non-primitive types by `const reference`

 4) write add/subtract/divide/multiply member functions for each type that take your 3 UDTs
        These are in addition to your member functions that take primitives
        for example, `IntType::divide(const DoubleType& dt);`

- [ ] revise